Public/Src/Tools/Execution.Analyzer/Analyzers.Core/XLGPlusPlus/XldbProtobufExtensions.cs

```diff
@@ -558,5 +548,479 @@ public static Xldb.Fingerprint ToFingerprint(this Fingerprint fingerprint)
                 Bytes = Google.Protobuf.ByteString.CopyFrom(fingerprint.ToByteArray())
             };
         }
+
+        /// <nodoc />
+        public static Xldb.DirectoryArtifact ToDirectoryArtifact(this DirectoryArtifact artifact, PathTable pathTable)
+        {
+            return new Xldb.DirectoryArtifact()
+            {
+                IsValid = artifact.IsValid,
+                Path = artifact.Path.ToAbsolutePath(pathTable),
+                PartialSealID = artifact.PartialSealId,
+                IsSharedOpaque = artifact.IsSharedOpaque
+            };
+        }
+
+        /// <nodoc />
+        public static Xldb.PipTable ToPipTable(this PipTable pipTable)
+        {
+            var xldbPipTable = new Xldb.PipTable
+            {
+                IsDisposed = pipTable.IsDisposed,
+                Reads = pipTable.Reads,
+                Writes = pipTable.Writes,
+                Count = pipTable.Count,
+                PageStreamsCount = pipTable.PageStreamsCount,
+                Size = pipTable.Size,
+                WritesMilliseconds = pipTable.WritesMilliseconds,
+                ReadsMilliseconds = pipTable.ReadsMilliseconds,
+                Used = pipTable.Used,
+                Alive = pipTable.Alive
+            };
+
+            xldbPipTable.StableKeys.AddRange(pipTable.StableKeys.Select(stableKey => stableKey.Value));
+            xldbPipTable.Keys.AddRange(pipTable.Keys.Select(key => key.Value));
+            xldbPipTable.DeserializationContexts.AddRange(pipTable.DeserializationContexts.Select(
+                context => new PipDeserializationContext() { Key = (PipQueryContext)context.Key, Value = context.Value }));
+            return xldbPipTable;
+        }
+
+        /// <nodoc />
+        public static Xldb.PipData ToPipData(this PipData pipData)

```

Comment:
>Xldb.PipData  [](start = 22, length = 13)

PipData is a bit of an internal implementation as well. We are abstracting file paths as strings. Should we do the same for PipData?